src/code/middleware/state-save.js

```diff
@@ -44,6 +52,21 @@ export default () => store => next => action => {
   return result;
 };
 
+const getCurrentChallenge = function(nextState) {
+  const {routeSpec, authoring, gems} = nextState;
+  if (!authoring) {
+    return null;
+  } else if (routeSpec && routeSpec.level !== undefined) {
+    return {
+      level: routeSpec.level,
+      mission: routeSpec.mission,
+      challenge: routeSpec.challenge
+    };
+  } else {
+    return progressUtils.getCurrentChallengeFromGems(authoring, gems);
```

Comment:
This is the only piece that confused me a bit, the idea that this piece of state contains either their current challenge or their next available challenge. Then again, any time they go to the home screen, they're directed to the next available challenge anyway, so this should be accurate, right?

Reply:
The idea is that the teachers will always want to see where a student is in the Report. If a student has finished a mission and goes back to the home room, then they don't have any current challenge in their state. So instead we use their gem state and work out what their next challenge is, using a pre-existing method. This *might* not end up being the one they chose to do next, as they could chose another challenge in the mission, or even a past mission, but it's our best guess as to where they are.

Alternatively, we could show what challenge they just finished (we'd have to add some stuff for that), but that seems less preferable, as it would look on the table that they are still working on that challenge (like maybe repeating it for a new gem).

Does that make sense?